Implement the capability to process chemical model NETCDF output and UFS-Aerosols model output

zhanglikate opened the pull request:
The previous UPP can only process the NEMSIO format for chemical output of GEFS-Aerosols. The updates has implemented the capability to read the chemical model NETCDF output for both GEFS-Aerosols (by turning on gocart_on) and UFS-Aerosols (by turning on nasa_on).
Also, re-assign the 6 AOD from GFS from 600-605 (rdaod) and Update the look-up table for UFS-Aerosols (nasa_on).

zhanglikate committed:
Implement the capability to read the chemical model NETCDF output for both GEFS-Aerosols (gocart_on) and UFS-Aerosols (nasa_on).

zhanglikate committed:
Merge branch 'develop' into ufs-aerosols-update

zhanglikate committed:
Merge remote-tracking branch 'upstream/develop' into ufs-aerosols-update

WenMeng-NOAA commented:
@zhanglikate Can you provide me model history files for testing new aerosol products? Thanks!

zhanglikate commented:
> @zhanglikate Can you provide me model history files for testing new aerosol products? Thanks!

@WenMeng-NOAA  You can find the netcdf output with chemical fields at: /scratch2/NCEPDEV/naqfc/Kate.Zhang/fv3gfs/comrot/NP8C_C96_ATM/gfs.20160715-nofb/00/atmos

Thanks.

WenMeng-NOAA commented:
@zhanglikate My test is at /scratch1/NCEPDEV/stmp2/Wen.Meng/post_gfs_2016071500. Can you verify if it is consistent with your test?

WenMeng-NOAA commented on sorc/ncep_post.fd/INITPOST_GFS_NEMS_MPIIO.f:
```diff
@@ -1336,16 +1336,6 @@ SUBROUTINE INITPOST_GFS_NEMS_MPIIO(iostatusAER)
 ! GFS output dust in nemsio (GOCART)
         dustcb=0.0
         dustallcb=0.0
-        do n=1,nbin_du
```
@zhanglikate I would like to confirm that initializing variables dust, salt, soot, waso, suso, pp25 and pp10 are removed as ALLOCATE_ALL is functioning that.

zhanglikate replied:
Yes. They are all defined and initialized ALLOCATE_ALL. Thanks.

WenMeng-NOAA commented:
@zhanglikate My test is at /scratch1/NCEPDEV/stmp2/Wen.Meng/post_gfs_2016071500. Can you verify if it is consistent with your test?

zhanglikate commented:
> @zhanglikate My test is at /scratch1/NCEPDEV/stmp2/Wen.Meng/post_gfs_2016071500. Can you verify if it is consistent with your test?

I am wondering how can I compare it? This is my test: /scratch1/NCEPDEV/stmp2/Kate.Zhang/post_gfs_2016071500

WenMeng-NOAA commented:
> > @zhanglikate My test is at /scratch1/NCEPDEV/stmp2/Wen.Meng/post_gfs_2016071500. Can you verify if it is consistent with your test?
> 
> I am wondering how can I compare it? This is my test: /scratch1/NCEPDEV/stmp2/Kate.Zhang/post_gfs_2016071500

@zhanglikate It looks the master file from UPP were not archived in your test. Can you take a look at the mast file from my test  GFSPRS.GrbF12 if all expected variables (postcntrl_ufs_aerosol.xml) are there?

zhanglikate commented:
@WenMeng-NOAA How can I check the file of GFSPRS.GrbF12 ? Which tool that I can use to check it? Thanks. Or any way that I can archived the master file in my test?

WenMeng-NOAA commented:
> @WenMeng-NOAA How can I check the file of GFSPRS.GrbF12 ? Which tool that I can use to check it? Thanks. Or any way that I can archived the master file in my test?

@zhanglikate The GFSPRS.GrbF12 is grib2 file generated in UPP with your UPP control file postcntrl_ufs_aerosol.xml. You might use wgrib2 to dump it.

zhanglikate commented:
> > @WenMeng-NOAA How can I check the file of GFSPRS.GrbF12 ? Which tool that I can use to check it? Thanks. Or any way that I can archived the master file in my test?
> 
> @zhanglikate The GFSPRS.GrbF12 is grib2 file generated in UPP with your UPP control file postcntrl_ufs_aerosol.xml. You might use wgrib2 to dump it.

@WenMeng-NOAA What is the resolution? 0.25, 0.5 or 1?

zhanglikate commented:
> > @WenMeng-NOAA How can I check the file of GFSPRS.GrbF12 ? Which tool that I can use to check it? Thanks. Or any way that I can archived the master file in my test?
> 
> @zhanglikate The GFSPRS.GrbF12 is grib2 file generated in UPP with your UPP control file postcntrl_ufs_aerosol.xml. You might use wgrib2 to dump it.

@WenMeng-NOAA I have the master output at: /scratch1/NCEPDEV/stmp2/Kate.Zhang/gfs.2016071500 , I just converted them to Netcdf and compared with your results, for the chemical related fields, they are identical. Thanks.

zhanglikate committed:
Revert as the develop branch ( global_1x1_paramlist_g2 and  scripts/exgfs_atmos_nceppost.sh)
Clean the unused dust variables in FDLVL.f and MDL2P.f

lipan-NOAA commented:
@zhanglikate is there any reason for this revert (3f43201) ?

WenMeng-NOAA commented:
> @zhanglikate is there any reason for this revert ([3f43201](https://github.com/NOAA-EMC/UPP/commit/3f43201ee0c43184b94e53187e24f97940b68ac6)) ?

@lipan-NOAA I suggested @zhanglikate to revert the changes because these two scripts are used for GFS post processing only. The changes could be added when the benchmarks for GFS V17 or GEFS V13 implementations are started.

WenMeng-NOAA commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5490,6 +5377,43 @@ SUBROUTINE CLDRAD
            datapd(1:iend-ista+1,1:jend-jsta+1,cfld)=GRID1(ista:iend,jsta:jend)
          endif
       ENDIF
+#if 0
```
@zhanglikate May I ask for the intention of the line "#if 0" and "endif"?

zhanglikate replied:
It means that the parts between  "#if 0" and "#endif" are commented out.

WenMeng-NOAA replied:
@zhanglikate I saw the warning message from compiling as:
```
/home/Wen.Meng/stmp2/kate/UPP/sorc/ncep_post.fd/CLDRAD.f(5380): warning #5117: Bad # preprocessor line
#if 0
-^
/home/Wen.Meng/stmp2/kate/UPP/sorc/ncep_post.fd/CLDRAD.f(5415): warning #5117: Bad # preprocessor line
#endif
```
The block is new code for outputting two new variables. You commented it out  temporarily?

zhanglikate commented:
Yes. I think we have to cancel the part by commenting out now. Thanks.

WenMeng-NOAA commented:
@zhanglikate Can you add change logs with your name at the beginning section of all fortran files you modified?

WenMeng-NOAA commented on sorc/ncep_post.fd/FDLVL.f:
```diff
@@ -812,6 +789,7 @@ SUBROUTINE FDLVL_UV(ITYPE,NFD,HTFD,UFD,VFD)
 !> 2017-06-01 | Y Mao        | Add FD levels for GTG(EDPARM CATEDR MWTURB) and allow levels input from control file
 !> 2019-09-25 | Y Mao        | Seperate mass from UV allow array of mass input to interpolate multiple fields with the same levels at one time. Dust=> AERFD can be processed when NIN=NBIN_DU
 !> 2020-11-10 | Jesse Meng   | Use UPP_PHYSICS module
+!> 2022-09-22 | Li(Kate) Zhang   | Remove Dust=> AERFD
```
@zhanglikate Should this change log be moved to after line 4?

zhanglikate replied:
@WenMeng-NOAA  Which line 4?

WenMeng-NOAA replied:
@zhanglikate Sorry for typo. Line 41.

WenMeng-NOAA commented on sorc/ncep_post.fd/MDL2P.f:
```diff
@@ -339,11 +338,11 @@ SUBROUTINE MDL2P(iostatusD3D)
-                     IF(DUST(I,J,1,K) < SPVAL) DUSTSL(I,J,K) = DUST(I,J,1,K)
-                   ENDDO
-                 endif
+!                 if (gocart_on) then
```
@zhanglikate Since these are legacy code, could they be deleted instead of commented out?

zhanglikate replied:
@WenMeng-NOAA  Removed. Thanks.

WenMeng-NOAA commented on sorc/ncep_post.fd/MDL2P.f:
```diff
@@ -501,12 +500,12 @@ SUBROUTINE MDL2P(iostatusD3D)
-                     DUSTSL(I,J,K) = DUST(I,J,LL,K) + (DUST(I,J,LL,K)-DUST(I,J,LL-1,K))*FACT
-                   ENDDO
-                 endif
+!                 if (gocart_on) then
```
@zhanglikate Since these are legacy code, could they be deleted instead of commented out?

zhanglikate replied:
@WenMeng-NOAA  Removed. Thanks.

WenMeng-NOAA commented on sorc/ncep_post.fd/MDL2P.f:
```diff
@@ -2062,130 +2061,7 @@ SUBROUTINE MDL2P(iostatusD3D)
              endif
           ENDIF
          ENDIF
-         if (gocart_on) then
```
@zhanglikate Since the output of UPP ID 439 to 442 in this routine, could you update post_avblflds.xml for removing their definitions?

zhanglikate replied:
@WenMeng-NOAA  Removed. Thanks.

lipan-NOAA reviewed (approved):
I looked through changes and tested it on Hera. It looks fine for me

zhanglikate committed:
1. Delete the commented out part
2. Re-assign the 6 AOD from GFS from 600-605 (rdaod).
3. Update the look-up table for UFS-Aerosols (nasa_on).

WenMeng-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -1105,136 +1102,6 @@ SUBROUTINE MISCLN
               ENDIF
             ENDIF
 !
-!  ADD FD LEVEL DUST/ASH (GOCART)
```
@zhanglikate You remove AERFD output (601-605) in this routine. Do you think the lines 839 to 853 can be removed too?
 ```
           IF (IGET(601)>0) THEN
            IF (LVLS(IFD,IGET(601))>1) ITYPEFDLVL(IFD)=2
           ENDIF
           IF (IGET(602)>0) THEN
            IF (LVLS(IFD,IGET(602))>1) ITYPEFDLVL(IFD)=2
           ENDIF
           IF (IGET(603)>0) THEN
            IF (LVLS(IFD,IGET(603))>1) ITYPEFDLVL(IFD)=2
           ENDIF
           IF (IGET(604)>0) THEN
            IF (LVLS(IFD,IGET(604))>1) ITYPEFDLVL(IFD)=2
           ENDIF
           IF (IGET(605)>0) THEN
            IF (LVLS(IFD,IGET(605))>1) ITYPEFDLVL(IFD)=2
           ENDIF
```

zhanglikate replied:
@WenMeng-NOAA  Yes. Updated.

lipan-NOAA commented:
@zhanglikate did you update the NASA lookup tables using Sarah's code ?

zhanglikate commented:
@lipan-NOAA  Yes. I am using the code you sent to me to update it, which I think it is from Sarah.

zhanglikate committed:
Remove DUST FLUX OF 601-605

WenMeng-NOAA commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5186,100 +5189,6 @@ SUBROUTINE CLDRAD
       ENDIF         ! END OF LAEROPT IF-BLOCK
 
-!! Multiply by 1.E-6 to revert these fields back
-      IF (IGET(659)>0) THEN
```
@zhanglikate With removal of the block of  output ID 659, do you think variable duem can be removed from this routine, and relevant routines ALLOCATE_ALL.f,  DEALLOCATE.f, INITPOST_GFS_NEMS_MPIIO.f, MPI_FIRST.f, VRBLS3D_mod.f, and parm/post_avblflds.xml?

zhanglikate replied:
@WenMeng-NOAA  I did not remove 659 for duem,  just remove the duplicated output of 659 in CLDRAD.f . We do need to output duem using 659.

WenMeng-NOAA replied:
@zhanglikate Got it. Thanks!

WenMeng-NOAA commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5186,100 +5189,6 @@ SUBROUTINE CLDRAD
-      ENDIF
-      
-!! Multiply by 1.E-6 to revert these fields back
-      IF (IGET(667)>0) THEN
```
@zhanglikate Do you think variable bcem can be cleaned up for the same reason above?

zhanglikate replied:
@WenMeng-NOAA  No.  As I mentioned above, we do need to output bcem using 667.

WenMeng-NOAA commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5186,100 +5189,6 @@ SUBROUTINE CLDRAD
-         endif
-      ENDIF
-
-      IF (IGET(660)>0) THEN
```
@zhanglikate Do you think variable dusd can be cleaned up for the same reason above?

zhanglikate replied:
@WenMeng-NOAA No. As I mentioned above, we do need to output dusd using 660.

WenMeng-NOAA commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5186,100 +5189,6 @@ SUBROUTINE CLDRAD
-         endif
-      ENDIF
-      
-      IF (IGET(699)>0) THEN
```
@zhanglikate Do you think variable maod can be cleaned up for the same reason above?

zhanglikate replied:
@WenMeng-NOAA No. As I mentioned above, we do need to output maod using 699.

WenMeng-NOAA commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5186,100 +5189,6 @@ SUBROUTINE CLDRAD
-      ENDIF
-!! ADD DUST DRY DEPOSITION FLUXES (kg/m2/sec)
-!
-!      IF (IGET(661)>0) THEN
```
@zhanglikate Do you think variable dudp can be cleaned up for the same reason above?

zhanglikate replied:
@WenMeng-NOAA No. As I mentioned above, we do need to output dudp using 661.

WenMeng-NOAA commented on sorc/ncep_post.fd/CLDRAD.f:
```diff
@@ -5311,26 +5220,6 @@ SUBROUTINE CLDRAD
       ENDIF
       
-!! ADD DUST WET DEPOSITION FLUXES (kg/m2/sec)
-!      IF (IGET(662)>0) THEN
```
@zhanglikate Do you think variable duwt can be cleaned up for the same reason above?

zhanglikate replied:
@WenMeng-NOAA No. As I mentioned above, we do need to output duwt using 662.

WenMeng-NOAA commented on sorc/ncep_post.fd/MDL2P.f:
```diff
@@ -57,8 +58,9 @@ SUBROUTINE MDL2P(iostatusD3D)
                             TD3D, IFHR, IFMIN, IM, JM, NBIN_DU, JSTA_2L,       &
-                            JEND_2U, LSM, d3d_on, gocart_on, ioform, NBIN_SM,  &
-                            imp_physics, ISTA, IEND, ISTA_M, IEND_M, ISTA_2L, IEND_2U
+                            JEND_2U, LSM, d3d_on, ioform, NBIN_SM,  &
```
@zhanglikate Do you think "DUSTSL" can be removed from following statements in this routine? 
```
!     REAL D3DSL(IM,JM,27),DUSTSL(IM,JM,NBIN_DU)
 REAL, allocatable  ::  D3DSL(:,:,:), DUSTSL(:,:,:), SMOKESL(:,:,:)
if(allocated(dustsl))  deallocate(dustsl)
```

zhanglikate replied:
@WenMeng-NOAA  Updated and pushed. Thanks.

zhanglikate committed:
Remove DUSTSL

zhanglikate committed:
Remove DUSTSL

zhanglikate committed:
Merge remote-tracking branch 'upstream/develop' into ufs-aerosols-update

WenMeng-NOAA commented on sorc/ncep_post.fd/MISCLN.f:
```diff
@@ -839,27 +836,10 @@ SUBROUTINE MISCLN
 !         print *,'call FDLVL with ITYPEFDLVL: ', ITYPEFDLVL,'for tmp,lvls=',LVLS(1:15,iget(59)), &
 !          'grib2tmp lvs=',LVLS(1:15,iget(586))
 
-         CALL FDLVL(ITYPEFDLVL,T7D,Q7D,U7D,V6D,P7D,ICINGFD,AERFD)
```
@zhanglikate This calling is needed for calculating other flight level fields. You may remove "AERFD" in this statement.

zhanglikate replied:
@WenMeng-NOAA  Updated.

WenMeng-NOAA commented on sorc/ncep_post.fd/FDLVL.f:
```diff
@@ -39,13 +39,13 @@
+!> 2022-09-22 | Li(Kate) Zhang   | Remove Dust=> AERFD
 !>
 !> @author Russ Treadon W/NP2 @date 1992-12-22
       SUBROUTINE FDLVL(ITYPE,TFD,QFD,UFD,VFD,PFD,ICINGFD,AERFD)
```
@zhanglikate You may remove "AERFD" in this line, line 68, and lines 101-111.

zhanglikate replied:
@WenMeng-NOAA  Updated.

zhanglikate committed:
Remove AERFD

WenMeng-NOAA commented:
The UPP RT tests were completed on WCOSS2, Hera and Orion. No baseline updates are needed with this PR.

WenMeng-NOAA commented:
@lipan-NOAA, @JianpingHuang-NOAA and @AnningCheng-NOAA Please let me know if any unexpected results are found from your testing for GEFS aerosol, AQM, and GFSV17 MERRA aerosol. Thanks!

lipan-NOAA commented:
@WenMeng-NOAA Could you point me the directory of your test on WCOSS2 ?

WenMeng-NOAA commented:
> @WenMeng-NOAA Could you point me the directory of your test on WCOSS2 ?

@lipan-NOAA Please see on cactus:

- the test with current develop branch: /u/wen.meng/ptmp/post_fv3chem_2022042400-before
- the test with Kate's branch: /u/wen.meng/ptmp/post_fv3chem_2022042400-after

The two tests use the same model history files provided from you. Thanks!

AnningCheng-NOAA commented:
The modifications related to MERRA2 (rdaod=T) looks fine and passed my tests. Approved.
